File: internal/blueprint/parse.go
package blueprint

import (
	"fmt"
	"os"
	"strings"
)

// debug is a local wrapper around tool package debug functionality
func debug(format string, args ...interface{}) {
	// We'll use fmt.Fprintf to stderr for now since we can't easily access tool.debug
	// This is a simple implementation that will work for our debugging needs
	if true { // Always log for now during debugging
		fmt.Fprintf(os.Stderr, "[Studio MCP Blueprint] "+format+"\n", args...)
	}
}

// FromArgs creates a new Blueprint from command arguments using tokenization
func FromArgs(args []string) (*Blueprint, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("cannot create blueprint: no command provided")
	}

	if strings.TrimSpace(args[0]) == "" {
		return nil, fmt.Errorf("cannot create blueprint: empty command provided")
	}

	// Debug logging: log the exact arguments received
	debug("FromArgs called with %d arguments:", len(args))
	for i, arg := range args {
		debug("  args[%d]: %q", i, arg)
	}

	bp := &Blueprint{
		BaseCommand: args[0],
		ShellWords:  make([][]Token, len(args)),
	}

	// Tokenize each shell word
	for i, arg := range args {
		tokens := tokenizeShellWord(arg)
		bp.ShellWords[i] = tokens
		debug("  shellword[%d] %q -> %d tokens", i, arg, len(tokens))
		for j, token := range tokens {
			debug("    token[%d]: %T %q", j, token, token.String())
		}
	}

	return bp, nil
}

// tokenizeShellWord tokenizes a single shell word into tokens
func tokenizeShellWord(word string) []Token {
	// Parse mixed content
	tokens := []Token{}
	pos := 0

	for pos < len(word) {
		templateStart := findNextTemplate(word, pos)
		if templateStart == nil {
			// No more templates, add remaining text
			if pos < len(word) {
				tokens = append(tokens, TextToken{Value: word[pos:]})
			}
			break
		}

		// Add text before template
		if templateStart.Start > pos {
			tokens = append(tokens, TextToken{Value: word[pos:templateStart.Start]})
		}

		// Parse template
		templateText := word[templateStart.Start:templateStart.End]
		if token := parseField(templateText); token != nil {
			tokens = append(tokens, token)
		} else {
			tokens = append(tokens, TextToken{Value: templateText})
		}

		pos = templateStart.End
	}

	// Ensure we always return at least one token
	if len(tokens) == 0 {
		tokens = append(tokens, TextToken{Value: word})
	}

	return tokens
}

// templateMatch represents a found template in the text
type templateMatch struct {
	Start        int
	End          int
	Type         string
	EndMarker    string
	MarkerLength int
}

// findNextTemplate finds the next template starting from the given position
func findNextTemplate(word string, startPos int) *templateMatch {
	remaining := word[startPos:]

	// Find possible starts
	startDouble := strings.Index(remaining, "{{")
	startSingle := strings.Index(remaining, "{")
	startOptional := strings.Index(remaining, "[")

	// If single coincides with double, prefer double
	if startSingle != -1 && startDouble != -1 && startSingle == startDouble {
		startSingle = -1
	}

	// Choose earliest non -1 among double, single, optional
	nextStart := -1
	templateType := ""
	endMarker := ""
	markerLength := 0

	candidates := []struct {
		idx  int
		typ  string
		end string
		len  int
	}{
		{startDouble, "required_double", "}}", 2},
		{startSingle, "required_single", "}", 1},
		{startOptional, "optional", "]", 1},
	}

	for _, c := range candidates {
		if c.idx != -1 && (nextStart == -1 || c.idx < nextStart) {
			nextStart = c.idx
			templateType = c.typ
			endMarker = c.end
			markerLength = c.len
		}
	}

	if nextStart == -1 {
		return nil // No templates found
	}

	absoluteStart := startPos + nextStart

	// Find template end
	endIndex := strings.Index(remaining[nextStart:], endMarker)
	if endIndex == -1 {
		// Malformed template - treat rest as text by returning no match
		return nil
	}

	absoluteEnd := absoluteStart + endIndex + markerLength

	return &templateMatch{
		Start:        absoluteStart,
		End:          absoluteEnd,
		Type:         templateType,
		EndMarker:    endMarker,
		MarkerLength: markerLength,
	}
}

// parseField parses a field enclosed in {{ }}, { } or [ ]
func parseField(field string) Token {
	var content string
	var required bool

	// Determine field type and extract content
	if strings.HasPrefix(field, "{{") && strings.HasSuffix(field, "}}") {
		content = field[2 : len(field)-2] // Remove {{ }}
		required = true
	} else if strings.HasPrefix(field, "{") && strings.HasSuffix(field, "}") {
		content = field[1 : len(field)-1] // Remove { }
		required = true
	} else if strings.HasPrefix(field, "[") && strings.HasSuffix(field, "]") {
		content = field[1 : len(field)-1] // Remove [ ]
		required = false
	} else {
		return nil // Not a valid field
	}

	// Parse content for name, description, and modifiers
	var name, description string
	isArray := false
	var originalFlag string

	// Check for description (split on #)
	parts := strings.SplitN(content, "#", 2)
	name = strings.TrimSpace(parts[0])

	// If name is empty, this is not a valid field (e.g., {{}} or {})
	if name == "" {
		return nil
	}

	if len(parts) > 1 {
		description = strings.TrimSpace(parts[1])
	}

	// Check for array notation (...)
	if strings.HasSuffix(name, "...") {
		isArray = true
		name = strings.TrimSuffix(name, "...")
		name = strings.TrimSpace(name)
	}

	// Check for boolean flag (starts with - or --)
	if !required && (strings.HasPrefix(name, "-") || strings.HasPrefix(name, "--")) {
		originalFlag = name
		name = strings.TrimLeft(name, "-")
		if description == "" {
			description = fmt.Sprintf("Enable %s flag", originalFlag)
		}
	}

	return FieldToken{
		Name:         name,
		Description:  description,
		Required:     required,
		IsArray:      isArray,
		OriginalFlag: originalFlag,
	}
}

File: src/blueprint/parse.ts
import { Blueprint, Token, TextToken, FieldToken } from './types.js';

/**
 * Template match represents a found template in the text
 */
interface TemplateMatch {
  start: number;
  end: number;
  type: string;
  endMarker: string;
  markerLength: number;
}

/**
 * Creates a new Blueprint from command arguments using tokenization
 */
export function fromArgs(args: string[]): Blueprint {
  if (args.length === 0) {
    throw new Error('cannot create blueprint: no command provided');
  }

  if (args[0].trim() === '') {
    throw new Error('cannot create blueprint: empty command provided');
  }

  const bp = new Blueprint(args[0], new Array(args.length));

  // Tokenize each shell word
  for (let i = 0; i < args.length; i++) {
    bp.shellWords[i] = tokenizeShellWord(args[i]);
  }

  return bp;
}

/**
 * Tokenizes a single shell word into tokens
 */
function tokenizeShellWord(word: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;

  while (pos < word.length) {
    const templateStart = findNextTemplate(word, pos);
    if (!templateStart) {
      // No more templates, add remaining text
      if (pos < word.length) {
        tokens.push(new TextToken(word.slice(pos)));
      }
      break;
    }

    // Add text before template
    if (templateStart.start > pos) {
      tokens.push(new TextToken(word.slice(pos, templateStart.start)));
    }

    // Parse template
    const templateText = word.slice(templateStart.start, templateStart.end);
    const token = parseField(templateText);
    if (token) {
      tokens.push(token);
    } else {
      tokens.push(new TextToken(templateText));
    }

    pos = templateStart.end;
  }

  // Ensure we always return at least one token
  if (tokens.length === 0) {
    tokens.push(new TextToken(word));
  }

  return tokens;
}

/**
 * Finds the next template starting from the given position
 */
function findNextTemplate(word: string, startPos: number): TemplateMatch | null {
  const remaining = word.slice(startPos);

  // Find possible starts
  let startDouble = remaining.indexOf('{{');
  let startSingle = remaining.indexOf('{');
  const startOptional = remaining.indexOf('[');

  // If single coincides with double, prefer double
  if (startSingle !== -1 && startDouble !== -1 && startSingle === startDouble) {
    startSingle = -1;
  }

  // Choose earliest non -1 among double, single, optional
  let nextStart = -1;
  let templateType = '';
  let endMarker = '';
  let markerLength = 0;

  const candidates = [
    { idx: startDouble, typ: 'required_double', end: '}}', len: 2 },
    { idx: startSingle, typ: 'required_single', end: '}', len: 1 },
    { idx: startOptional, typ: 'optional', end: ']', len: 1 },
  ];

  for (const c of candidates) {
    if (c.idx !== -1 && (nextStart === -1 || c.idx < nextStart)) {
      nextStart = c.idx;
      templateType = c.typ;
      endMarker = c.end;
      markerLength = c.len;
    }
  }

  if (nextStart === -1) {
    return null; // No templates found
  }

  const absoluteStart = startPos + nextStart;

  // Find template end
  const endIndex = remaining.slice(nextStart).indexOf(endMarker);
  if (endIndex === -1) {
    // Malformed template - treat rest as text by returning no match
    return null;
  }

  const absoluteEnd = absoluteStart + endIndex + markerLength;

  return {
    start: absoluteStart,
    end: absoluteEnd,
    type: templateType,
    endMarker,
    markerLength,
  };
}

/**
 * Parses a field enclosed in {{ }}, { } or [ ]
 */
function parseField(field: string): Token | null {
  let content: string;
  let required: boolean;

  // Determine field type and extract content
  if (field.startsWith('{{') && field.endsWith('}}')) {
    content = field.slice(2, -2); // Remove {{ }}
    required = true;
  } else if (field.startsWith('{') && field.endsWith('}')) {
    content = field.slice(1, -1); // Remove { }
    required = true;
  } else if (field.startsWith('[') && field.endsWith(']')) {
    content = field.slice(1, -1); // Remove [ ]
    required = false;
  } else {
    return null; // Not a valid field
  }

  // Parse content for name, description, and modifiers
  let name: string;
  let description = '';
  let isArray = false;
  let originalFlag = '';

  // Check for description (split on #)
  const parts = content.split('#');
  name = parts[0].trim();

  // If name is empty, this is not a valid field (e.g., {{}} or {})
  if (name === '') {
    return null;
  }

  if (parts.length > 1) {
    description = parts.slice(1).join('#').trim();
  }

  // Check for array notation (...)
  if (name.endsWith('...')) {
    isArray = true;
    name = name.slice(0, -3).trim();
  }

  // Check for boolean flag (starts with - or --)
  if (!required && (name.startsWith('-') || name.startsWith('--'))) {
    originalFlag = name;
    name = name.replace(/^-+/, '');
    if (description === '') {
      description = `Enable ${originalFlag} flag`;
    }
  }

  return new FieldToken(name, description, required, isArray, originalFlag);
}
